src/hasSubObject.test.ts

```diff
@@ -0,0 +1,158 @@
+import { hasSubObject } from "./hasSubObject";
+import { pipe } from "./pipe";
+
+describe("data first", () => {
+  test("works with empty sub-object", () => {
+    expect(hasSubObject({ a: 1, b: "b", c: 3 }, {})).toBe(true);
+    expect(hasSubObject({}, {})).toBe(true);
+  });
+
+  test("works with primitives", () => {
+    expect(hasSubObject({ a: 1, b: "b", c: 3 }, { a: 1, b: "b" })).toBe(true);
+    expect(hasSubObject({ a: 1, b: "c", c: 3 }, { a: 1, b: "b" })).toBe(false);
+    expect(hasSubObject({ a: 2, b: "b", c: 3 }, { a: 1, b: "b" })).toBe(false);
+  });
+
+  test("works with deep objects", () => {
+    expect(hasSubObject({ a: { b: 1, c: 2 } }, { a: { b: 1, c: 2 } })).toBe(
+      true,
+    );
+    expect(hasSubObject({ a: { b: 1, c: 2 } }, { a: { b: 1, c: 0 } })).toBe(
+      false,
+    );
+  });
+
+  test("checks for matching key", () => {
+    const data: { a?: undefined } = {};
+    expect(hasSubObject(data, { a: undefined })).toBe(false);
+  });
+
+  test("allows weird arguments", () => {
+    expect(hasSubObject(new Error("a"), { message: "a" })).toBe(true);
+    expect(hasSubObject(new Error("a"), { message: "b" })).toBe(false);
+
+    // Error has no enumerable properties, so this is true:
+    expect(hasSubObject(new Error("a"), new Error("b"))).toBe(true);
+
+    expect(hasSubObject(["a", "b"], ["a"])).toBe(true);
+    expect(hasSubObject(["a"], ["a", "b"])).toBe(false);
+
+    expect(hasSubObject("ab", "a")).toBe(true);
+    expect(hasSubObject("a", "ab")).toBe(false);
+  });
```

Comment:
**question:** should we have these as tests?

i think the Error tests are reasonable, less so the array and string ones

notably, there are no type errors with calling something like `hasSubObject(1, 2)` (which returns `true`)

Reply:
I don't think we should officially test anything except proper objects. I don't mind the type being loose enough to allow people to use it with weird edge-cases, but by testing it it means we support these cases, and can't break them in the future.